### PR TITLE
CORE-20522 Improve reconciler

### PR DESCRIPTION
**Pull request by williamvigorr3**
Improve the HostedIdentityReconciler, so it doesn't use the message bus RPC pattern anymore. Instead use a direct query to get the certificates from the Database and query the crypto worker using http RPC instead. I added an HTTP RPC endpoint to do this, to lookup public keys in the crypto worker.

Requires: https://github.com/corda/corda-api/pull/1659.

This is currently targeting 5.2 but we should change it to target 5.3, once we have merged 5.2 into 5.3.

I need to add some unit tests but before I do that it would be useful to get some feedback on the implementation.

Testing
----
1. Onboard an MGM and Notary members into Corda (running the combined worker with a Kafka message bus).
2. Stop the combined worker process
3. Delete the topic p2p.hosted.identities
4. Recreate the topic p2p.hosted.identities
5. Check data successfully written to Kafka
```

william.vigor@LDNM-WHD6VPFHQC corda-runtime-os % /opt/homebrew/opt/kafka/bin/kafka-topics --bootstrap-server localhost:9092 --delete  --topic p2p.hosted.identities
william.vigor@LDNM-WHD6VPFHQC corda-runtime-os % /opt/homebrew/opt/kafka/bin/kafka-topics --bootstrap-server localhost:9092 --create  --topic p2p.hosted.identities
WARNING: Due to limitations in metric names, topics with a period ('.') or underscore ('_') could collide. To avoid issues it is best to use either, but not both.
Created topic p2p.hosted.identities.
william.vigor@LDNM-WHD6VPFHQC corda-runtime-os % /opt/homebrew/opt/kafka/bin/kafka-console-consumer --bootstrap-server localhost:9092 --topic p2p.hosted.identities --from-beginning
corda�_{��_*e�{�U����<�R��� %��L��CN=MGM, O=ReRegistrationMultiCluster-54850110-ef58-4424-9c47-67632a51e83a, L=London, C=GBH998f6355-8b3d-40b1-b386-81e6a22198e6p2p�-----BEGIN CERTIFICATE-----
MIIDPjCCAaagAwIBAgIBBTANBgkqhkiG9w0BAQsFADBNMQswCQYDVQQGEwJVSzEP
MA0GA1UEAwwGcjMuY29tMS0wKwYDVQQLDCQ3MGVmMDlmMy0yZjdjLTRlMzgtODQy
My1lMzExNzYzMzcyMzAwHhcNMjQwNjEyMTAyMDQxWhcNMjQwNzEyMTAyMDQxWjB2
MQswCQYDVQQGEwJHQjEPMA0GA1UEBxMGTG9uZG9uMUgwRgYDVQQKEz9SZVJlZ2lz
dHJhdGlvbk11bHRpQ2x1c3Rlci01NDg1MDExMC1lZjU4LTQ0MjQtOWM0Ny02NzYz
MmE1MWU4M2ExDDAKBgNVBAMTA01HTTBZMBMGByqGSM49AgEGCCqGSM49AwEHA0IA
BKKKIU9ytbXcj9T3aPIoxdw1czgQAhRFbzVle3lpEh6lBK+2rHRlCDD48EhtcRG6
LssUbCH7qMwwtBXeqSjy0QGjSzBJMA4GA1UdDwEB/wQEAwIHgDA3BgNVHREBAf8E
LTArgiljb3JkYS1wMnAtZ2F0ZXdheS13b3JrZXIud3ZpZ29yLWNsdXN0ZXItMzAN
BgkqhkiG9w0BAQsFAAOCAYEANrPxlQeLV2JyphBVsyWM6X4/dA59mOReHz3Am+wu
eFH6AuuWS1DynS4idmJWh+qdk/nt8sKDLoiQDL5g0bHLptoYOhIiA9M5m9qAeywJ
3qljsa3+E7nIQ7kq6A1lzG8kK3YZDHDLGR88CT4V4ypU+pRRKZYzfa1nB9YZiCM2
tppvZxxCix7hMxNp9TEJ+YslFextGr/UEsWdn4hSFQt2XVQlNVmusVNs/af4Vzye
ZT3VVZQ7VMSmoszrIK4mn0UTdg7Wt/z3zUIdXfWofn1tsKWclW9g1wRs7XMnfRBi
6nkJdrf6bPOdAHC8ID4v7eU0paAkR4XofRXByvKHvvc3zr0hauT/7R+GCU86yFvU
Fr15MZPBEspcSMWxANzwOs2IZYET5SmDD19+XzcpS1GJIdbwYLxmc2GmopdX/K7g
z5iFO7tW5ZaiIadjucOLCogCvSi95B21hFH6MNEWTAI/SGSaQKlCfoRLFH5aBJd7
QqTt2BmSTfgSreOdTxTohCis
-----END CERTIFICATE-----
�-----BEGIN PUBLIC KEY-----
MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAE5iKgmIzGT2m1YjL+8fZShFBfPRmV
Fr30pVZDhna+hgfjWY+okELxfWck6ezpPe28s3bZMPlhlme3u5GWJJpqnA==
-----END PUBLIC KEY-----

corda�_{��_*e�{�U����<�R��� %��L��CN=Notary, O=ReRegistrationMultiCluster-54850110-ef58-4424-9c47-67632a51e83a, L=London, C=GBH998f6355-8b3d-40b1-b386-81e6a22198e6p2p�-----BEGIN CERTIFICATE-----
MIIDPjCCAaagAwIBAgIBBTANBgkqhkiG9w0BAQsFADBNMQswCQYDVQQGEwJVSzEP
MA0GA1UEAwwGcjMuY29tMS0wKwYDVQQLDCQ3MGVmMDlmMy0yZjdjLTRlMzgtODQy
My1lMzExNzYzMzcyMzAwHhcNMjQwNjEyMTAyMDQxWhcNMjQwNzEyMTAyMDQxWjB2
MQswCQYDVQQGEwJHQjEPMA0GA1UEBxMGTG9uZG9uMUgwRgYDVQQKEz9SZVJlZ2lz
dHJhdGlvbk11bHRpQ2x1c3Rlci01NDg1MDExMC1lZjU4LTQ0MjQtOWM0Ny02NzYz
MmE1MWU4M2ExDDAKBgNVBAMTA01HTTBZMBMGByqGSM49AgEGCCqGSM49AwEHA0IA
BKKKIU9ytbXcj9T3aPIoxdw1czgQAhRFbzVle3lpEh6lBK+2rHRlCDD48EhtcRG6
LssUbCH7qMwwtBXeqSjy0QGjSzBJMA4GA1UdDwEB/wQEAwIHgDA3BgNVHREBAf8E
LTArgiljb3JkYS1wMnAtZ2F0ZXdheS13b3JrZXIud3ZpZ29yLWNsdXN0ZXItMzAN
BgkqhkiG9w0BAQsFAAOCAYEANrPxlQeLV2JyphBVsyWM6X4/dA59mOReHz3Am+wu
eFH6AuuWS1DynS4idmJWh+qdk/nt8sKDLoiQDL5g0bHLptoYOhIiA9M5m9qAeywJ
3qljsa3+E7nIQ7kq6A1lzG8kK3YZDHDLGR88CT4V4ypU+pRRKZYzfa1nB9YZiCM2
tppvZxxCix7hMxNp9TEJ+YslFextGr/UEsWdn4hSFQt2XVQlNVmusVNs/af4Vzye
ZT3VVZQ7VMSmoszrIK4mn0UTdg7Wt/z3zUIdXfWofn1tsKWclW9g1wRs7XMnfRBi
6nkJdrf6bPOdAHC8ID4v7eU0paAkR4XofRXByvKHvvc3zr0hauT/7R+GCU86yFvU
Fr15MZPBEspcSMWxANzwOs2IZYET5SmDD19+XzcpS1GJIdbwYLxmc2GmopdX/K7g
z5iFO7tW5ZaiIadjucOLCogCvSi95B21hFH6MNEWTAI/SGSaQKlCfoRLFH5aBJd7
QqTt2BmSTfgSreOdTxTohCis
-----END CERTIFICATE-----
�-----BEGIN PUBLIC KEY-----
MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEoOcv2VCAO8feuSH2TwuiOXwAC1pc
h5D8Uy1vMxfKXGGyyFflOjv8nkY8+bk/o16yxeTWQ4caeuEWE1OK25yIww==
-----END PUBLIC KEY-----
```

### CODE REVIEW

**Review by kyriathar — COMMENTED** [review]
Had a quick pass. 

As a general thing, I am wondering whether we should be switching over to sync processing generally when it comes to this operation, i.e. polling for keys. I think right now we have both? 

But maybe that's too much of a change for 5.2, maybe we could not do the sync keys polling in 5.2 and then we are just left with using DB client directly for certificates and the change to streaming you made already brings us in an OK state with minimal changes for 5.2?

File: processors/db-processor/src/main/kotlin/net/corda/processors/db/internal/reconcile/db/HostedIdentityReconciler.kt [hunk 1]
```diff
@@ -181,23 +208,35 @@ class HostedIdentityReconciler(
...
     }
 
     private fun getSessionKey(
         tenantId: String,
         sessionKeyId: ShortHash,
     ): String {
-        return cryptoOpsClient.lookupKeysByIds(tenantId, listOf(sessionKeyId)).firstOrNull()
-            ?.toPem()
-            ?: throw CertificatesResourceNotFoundException("Can not find session key for $tenantId")
+        val cachedCertificate = cachedKeys.getIfPresent(KeyLookup(tenantId, sessionKeyId))
```
Comment (YashNabar) [inline processors/db-processor/src/main/kotlin/net/corda/processors/db/internal/reconcile/db/HostedIdentityReconciler.kt]: key?

File: components/crypto/crypto-client/src/main/kotlin/net/corda/crypto/client/ReconcilerCryptoOpsClient.kt [hunk 2]
```diff
@@ -0,0 +1,9 @@
+package net.corda.crypto.client
+
+import net.corda.crypto.core.ShortHash
+import net.corda.data.crypto.wire.CryptoSigningKey
+import net.corda.lifecycle.Lifecycle
+
+interface ReconcilerCryptoOpsClient: Lifecycle {
+    fun lookupKeysByIds(tenantId: String, keyIds: List<ShortHash>): List<CryptoSigningKey>
```
Comment (YashNabar) [inline components/crypto/crypto-client/src/main/kotlin/net/corda/crypto/client/ReconcilerCryptoOpsClient.kt]: Can we add some kdocs to this interface?

**Review by YashNabar — COMMENTED** [review]
Changes to the reconciler implementation look good to me overall. 
Do we want to reconsider the "reconciler" naming references? Since we generally want to move towards HTTP RPC anyway, those endpoints won't be reconciler-specific for long?

**Pull request closed without merging**